### PR TITLE
Add override support for Knative Services

### LINKED ISSUE #838: Add override support for Knative Services
Knative services don't support `konghq.com/override`.
Add support for Knative services for this.

See https://github.com/Kong/kubernetes-ingress-controller/issues/742 for more details.

### PR DESCRIPTION
**Pull request by rafaelfelix**
**What this PR does / why we need it**: Add override support for Knative Services

**Which issue this PR fixes**: fixes #838

**Special notes for your reviewer**:

### CODE REVIEW

**Review by mflendrich — CHANGES_REQUESTED** [review]
Hi @rafaelfelix,

Thanks for your contribution!

Could you please add a test under `parser_test.go` for cases where a Knative Ingress has, and doesn't have, an override annotation?

For reference, you can reuse `TestKnativeIngressAndPlugins`: create a test like this:
- Arrange: create a `NewFakeStore()`  with a `KnativeIngress` and a `KongIngress` therein,
- Act: `Build(logger, fakestore)`
- Assert: that the resulting KongState object has the expected overrides in place

**Comment by rafaelfelix** [comment]
hi @mflendrich! just pushed the tests as requested. let me know if there's any other case we gotta cover :) thanks

**Review by mflendrich — DISMISSED** [review]
Looks good to me!

@hbagdi , is there anything we've missed that needs attention? You're more familiar with the issue than I am. If it's good to go, please merge :fire:

File: internal/ingress/controller/parser/parser_test.go [hunk 1]
```diff
@@ -1732,6 +1732,175 @@ func TestKongProcessClasslessIngress(t *testing.T) {
...
+			},
+		}
+		kongIngresses := []*configurationv1.KongIngress{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "https-only",
+					Namespace: "foo-ns",
+					Annotations: map[string]string{
+						"networking.knative.dev/ingress.class": annotations.DefaultIngressClass,
+					},
```
Comment (hbagdi) [inline internal/ingress/controller/parser/parser_test.go:1782]: Why is this annotation added in here?
This annotation should only be interpreted on the Knative Ingress resource.

File: internal/ingress/controller/parser/parser_test.go [hunk 2]
```diff
@@ -1732,6 +1732,175 @@ func TestKongProcessClasslessIngress(t *testing.T) {
...
+				},
+			},
+		}
+		services := []*corev1.Service{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo-svc",
+					Namespace: "foo-ns",
+					Annotations: map[string]string{
+						"networking.knative.dev/ingress.class": annotations.DefaultIngressClass,
```
Comment (hbagdi) [inline internal/ingress/controller/parser/parser_test.go:1796]: Same question here.

File: internal/ingress/controller/parser/parser_test.go [hunk 3]
```diff
@@ -1732,6 +1732,175 @@ func TestKongProcessClasslessIngress(t *testing.T) {
...
+
+		assert.Equal(1, len(state.Services), "expected one knative service")
+		svc := state.Services[0]
+
+		assert.Equal(1, len(svc.Routes), "expected one route in knative service")
+		route := svc.Routes[0]
+
+		assert.NotEqual(kong.StringSlice("https"), route.Protocols)
+		assert.Nil(route.HTTPSRedirectStatusCode)
+	})
```
Comment (hbagdi) [inline internal/ingress/controller/parser/parser_test.go:1903]: Please add in at least one test which ensure that other annotations (`konghq.com/protocols`, `konghq.com/strip-path` etc)work too. I expect such annotations to be the common case over `konghq.com/override`.

File: internal/ingress/controller/parser/parser_test.go (hunk 1 above)
Comment (rafaelfelix) [reply internal/ingress/controller/parser/parser_test.go:1782]: Thanks for pointing it out. I probably copy-pasted from another section and forgot to remove it.

File: internal/ingress/controller/parser/parser_test.go (hunk 2 above)
Comment (rafaelfelix) [reply internal/ingress/controller/parser/parser_test.go:1796]: same reply as above

File: internal/ingress/controller/parser/parser_test.go (hunk 3 above)
Comment (rafaelfelix) [reply internal/ingress/controller/parser/parser_test.go:1891]: added as requested.

**Comment by rafaelfelix** [comment]
@hbagdi, @mflendrich please review again whenever you have time. Thanks

**Review by hbagdi — APPROVED** [review]

**Comment by hbagdi** [comment]
@rafaelfelix This needs to target the `next` branch. Could you please rebase your branch on top of next?

**Comment by rafaelfelix** [comment]
@hbagdi just rebased my branch with `next` and force-pushed. please review again when you have some time.

thank you

**Review by hbagdi — APPROVED** [review]
It is interesting to see how a single line patch can add this functionality now.

**Comment by hbagdi** [comment]
Thank you for your contribution. Please fill out the following form to claim your contributor swag:
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributor-t-shirt.

**Pull request merged**